renpy/common/00action_data.rpy

```diff
@@ -34,372 +35,333 @@ init -1600 python:
-            If not None, then this is the true value used.
-        `false_value`
-            If not None, then this is the false value used.
+        Warning : the provided values must not be duplicated, otherwise some values can never be set.
         """
 
-        return ToggleField(store, variable, true_value=true_value, false_value=false_value, kind="variable")
+        identity_fields = ()
+        equality_fields = ("values", "loop")
 
+        def __init__(self, values, *args, **kwargs):
+            super(Cycle, self).__init__(*args, **kwargs)
 
-    @renpy.pure
-    class ToggleDict(Action, FieldEquality):
-        """
-        :doc: data_action
-
-        Toggles the value of `key` in `dict`. It also works on
-        lists, in which case `key` is the index of the value to toggle.
-        Toggling means to invert the value when the action is performed.
+            if kwargs.get("reverse", False):
+                values = values[::-1]
+                # the reversed builtin returns a non-indexable iterator
+                # and we only support sequences anyway - iterators are not picklable
+            self.loop = kwargs.get("loop", True)
 
-        `true_value`
-            If not None, then this is the true value used.
-        `false_value`
-            If not None, then this is the false value used.
-        """
+            self.values = renpy.easy.to_tuple(values)
 
-        identity_fields = [ "dict", ]
-        equality_fields = [ "key", "true_value", "false_value" ]
+        def value_to_set(self):
+            values = self.values
+            value = self.current_value()
 
-        def __init__(self, dict, key, true_value=None, false_value=None):
-            self.dict = dict
-            self.key = key
-            self.true_value = true_value
-            self.false_value = false_value
+            if value in values:
+                idx = values.index(value) + 1
 
-        def __call__(self):
-            value = self.dict[self.key]
+                lnv = len(values)
+                if self.loop and idx >= lnv:
+                    idx -= lnv
 
-            if self.true_value is not None:
-                value = (value == self.true_value)
-
-            value = not value
+            else:
+                idx = 0
 
-            if self.true_value is not None:
-                if value:
-                    value = self.true_value
-                else:
-                    value = self.false_value
+            return values[idx]
 
-            self.dict[self.key] = value
-            renpy.restart_interaction()
+        def get_sensitive(self):
+            if self.loop:
+                return True
 
-        def get_selected(self):
             try:
-                rv = self.dict[self.key]
-            except (KeyError, IndexError):
+                value = self.current_value()
+            except (__LookupError, __NoCurrentScreen, __ScreenVariableNameError): # __FieldNotFound ?
                 return False
```

Comment:
I can't write as a suggestion because this covers deleted lines.
The test for `self.loop` should occur later to be able to handle cases where the loop is small (0 or 1-sized).

```py
            try:
                rv = self.dict[self.key]
            except (KeyError, IndexError):
                value = self.current_value()
            except (__LookupError, __NoCurrentScreen, __ScreenVariableNameError): # __FieldNotFound ?
                return False

            if self.loop:
                return len(self.values) > 1 or (len(self.values) == 1 and self.values[0] != rv)
```

Reply:
I changed my mind for the empty sequence case, because it's cheap in execution time.
It will still raise when the action is run of course.